From 3afac6d4cc0b772be4abf755dd063617e30196f3 Mon Sep 17 00:00:00 2001
From: Sailaja <sbandarupalli@habana.ai>
Date: Fri, 13 Nov 2020 19:50:53 +0200
Subject: [PATCH] [SW-24414] goya2: boot fit loading bug fix

https://gerrit.habana-labs.com/56778 patch caused regression in Goya2
For legacy+ boot flow, Arc images are loaded as part of boot FIT
and need to consider ARC0 and ARC1 ICCM,DCCM memories as valid load
addresses

Change-Id: Ifa162f289bd7b1c3b14eec2d32b00a36c457b575
Signed-off-by: Sailaja <sbandarupalli@habana.ai>
---

diff --git a/goya2/common.c b/goya2/common.c
index d89e957..782422b 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -520,5 +520,16 @@
 	    addr + len < HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE)
 		return true;
 
-	return preboot_sram_is_valid_address_range(addr, len);
+	if (preboot_sram_is_valid_address_range(addr, len))
+		return true;
+
+	if (addr >= ARC0_ICCM_BASE_ADDR &&
+	    addr + len < ARC0_ICCM_BASE_ADDR + ARC0_MEM_SIZE)
+		return true;
+
+	if (addr >= ARC1_ICCM_BASE_ADDR &&
+	    addr + len < ARC1_ICCM_BASE_ADDR + ARC1_MEM_SIZE)
+		return true;
+
+	return false;
 }
diff --git a/goya2/common.h b/goya2/common.h
index a70b17f..0df56d6 100644
--- a/goya2/common.h
+++ b/goya2/common.h
@@ -7,6 +7,7 @@
 
 #include <preboot.h>
 #include <goya2_blocks_linux_driver.h>
+#include <goya2/goya2.h>
 
 #define SRAM_SIZE			0x8000000	/* 128MB */
 #define HL_SP_SRAM_SIZE			0x10000
@@ -24,6 +25,8 @@
 
 #define ARC_RST_VALUE			0x110
 #define ARC_RST_VEC_ADDR		0x0
+#define ARC0_MEM_SIZE			(ARC0_ICCM_SIZE + ARC0_DCCM_SIZE)
+#define ARC1_MEM_SIZE			(ARC1_ICCM_SIZE + ARC1_DCCM_SIZE)
 
 #define HL_PCIE_LINK_RETRIES		50
 #define HL_INIT_SP_SRAM_RETRIES		50
